### PR TITLE
[dynamo] Fix TorchFunctionMode leak on graph break (#182317)

### LINKED ISSUE #182317: Dynamo leaks TorchFunctionMode onto stack across graph breaks
## Summary

When a `with TorchFunctionMode()` context manager spans a `__build_class__` graph break, dynamo's resume function leaves a mode on the C-level torch function mode stack. This silently poisons subsequent tests/code.

## Repro

```python
import torch
from torch.overrides import TorchFunctionMode

@torch.compile(backend="eager")
def repro():
    class A(TorchFunctionMode):
        def __torch_function__(self, *args, **kwargs):
            return -1

    with A():
        torch.tensor([1])

repro()
print(torch._C._len_torch_function_stack())  # prints 1, should be 0
```

## Investigation

Found via `test_overrides.py::TestTorchFunctionMode::test_factory_override` poisoning later tests in the same process. Bisected using subprocess isolation to confirm this single test leaks a mode.

The `__build_class__` call for the `TorchFunctionMode` subclass triggers a graph break. Dynamo exits the `with` context symbolically and generates a resume function that re-enters it. The resume function's prologue path enters `DisableTorchFunctionSubclass` and correctly exits it, but the `TorchFunctionMode` push/pop across the graph break boundary is not properly balanced.

## Impact

Any test or user code running after this in the same process will see unexpected torch function modes on the stack, causing incorrect dispatch behavior, assertion failures, or mysterious type errors.

cc @hameerabbasi @rgommers @ezyang @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @Lucaskabela @jataylo @azahed98

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* __->__ #183069

Fix #182317

When a TorchFunctionMode context manager spanned a graph break (e.g., from
__build_class__ or an untraceable function like sparse_coo_tensor), the
resume function's exception handler referenced ___prev_torch_function_mode_stack
-- a variable set by the output bytecode's side effects codegen. But this variable
lived in the *first* compilation's scope, not the resume function's scope, so the
LOAD_FAST in the except handler hit a KeyError. That KeyError masked the original
exception and left the mode on the C-level torch function mode stack, silently
poisoning subsequent code in the same process.

The fix: swap the except handler from set_torch_function_mode_stack(prev_stack)
(which needed the missing variable) to a simple _pop_torch_function_stack(). This
matches the push that the compiled graph output did, and does not require any
cross-scope variable access. The normal exit path was already correct -- the
original with-block's __exit__ pops the mode -- so only the exception path needed
fixing.

Also removes the skipIf decorator on test_factory_override in test_overrides.py,
since the root cause is now fixed.

Authored by Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

File: test/dynamo/test_modes.py [hunk 1]
```diff
@@ -893,6 +893,44 @@ def with_explicit_device(
...
+            class A(TorchFunctionMode):
+                def __torch_function__(self, *args, **kwargs):
+                    return -1
+
+            with A():
+                _x = torch.tensor([1])
+
+                # Second graph break inside the with block
+                class B:
+                    pass
```
Comment (williamwen42) [inline test/dynamo/test_modes.py]: nit: Could this just be `torch._dynamo.graph_break()`?

**Review by williamwen42 — COMMENTED** [review]
cc @mlazos since this is a semantically significant bytecode change.

**Comment by mlazos** [comment]
@claude does this handle the default device torch function mode correctly do you think? I'm curious if a simple pop is okay here, it it possible other side effects have changed the mode stack beyond just the push?

File: torch/_dynamo/symbolic_convert.py [hunk 2]
```diff
@@ -984,8 +984,6 @@ def _reconstruct_block_stack(
         cur_tx = cur_tx.parent
     for tx in reversed(all_txes):
         for b in tx.block_stack:
-            # Don't exit any modes we have entered,
```
Comment (mlazos) [inline torch/_dynamo/symbolic_convert.py:987]: I think this comment is still valid, right? the output bytecode will pop the mode accordingly

**Review by mlazos — COMMENTED** [review]
Looks good, left one comment about leaving a comment, but it's non-blocking. thanks for debugging this! This is a good simplification and bugfix.

**Review by mlazos — APPROVED** [review]
